# Descompunerea in valori singulare

Bun venit in laboratorul de metode numerice. Astazi ne vom ocupa de studiul valorilor singulare, precum si de decompunerea unei matrici cu ajutorul acestora.
Pentru inceput, trebuie sa trecem prin laboratorul dedicat acestei lectii, pentru a putea intelege fara mari probleme codul si micile "experimente" pe care le planuim sa le facem. Sa incepem!

Sa ne amintim de matricea X din laborator, a carei descompunere in valori singulare este deja facuta. Folosind aceasta matrice, vom putea compara in timp real rezultatul
algoritmilor pe care ii vom prezenta cu rezultatul matematic obtinut.

$$X = \begin{pmatrix} 1 & 0 & 1 \\ 0 & 1 & 0 \\ 0 & 1 & 1 \\ 0 & 1 & 0 \\ 1 & 1 & 0 \end{pmatrix}$$

Descompunerea acesteia in valori singulare ne-a dat:

$$X = \begin{pmatrix} \frac{\sqrt{30}}{15} & \frac{\sqrt{6}}{3} & 0 & 0 & -\frac{\sqrt{7}}{7} \\ \frac{\sqrt{30}}{15} & -\frac{\sqrt{6}}{6} & 0 & -\frac{\sqrt{2}}{2} & \frac{2\sqrt{7}}{7} \\ \frac{\sqrt{30}}{10} & 0 & \frac{\sqrt{2}}{2} & 0 & \frac{\sqrt{7}}{7} \\ \frac{\sqrt{30}}{15} & -\frac{\sqrt{6}}{6} & 0 & \frac{\sqrt{2}}{2} & 0 \\ \frac{\sqrt{30}}{10} & 0 & -\frac{\sqrt{2}}{2} & 0 & \frac{\sqrt{7}}{7} \end{pmatrix} * \begin{pmatrix} \sqrt{5} & 0 & 0 \\ 0 & \sqrt{2} & 0 \\ 0 & 0 & 1 \\ 0 & 0 & 0 \\ 0 & 0 &0 \end{pmatrix} * \begin{pmatrix} \frac{\sqrt{6}}{6} & \frac{\sqrt{6}}{3} & \frac{\sqrt{6}}{6} \\ \frac{\sqrt{3}}{3} & -\frac{\sqrt{3}}{3} & \frac{\sqrt{3}}{3} \\ -\frac{\sqrt{2}}{2} & 0 & \frac{\sqrt{2}}{2} \end{pmatrix}$$

Cu X<sub>[5x3]</sub> = U<sub>[5x5]</sub> * S<sub>[5x3]</sub> * V<sup>T</sup><sub>[3x3]</sub>

In primul rand, vom initializa matricea X, si vom forma matricile corespunzatoare A<sub>1</sub> si A<sub>2</sub>.

In [3]:
X = [1 0 1; 0 1 0; 0 1 1; 0 1 0; 1 1 0]

X =

   1   0   1
   0   1   0
   0   1   1
   0   1   0
   1   1   0



In [4]:
A1 = X * X'

A1 =

   2   0   1   0   1
   0   1   1   1   1
   1   1   2   1   1
   0   1   1   1   1
   1   1   1   1   2



In [7]:
A2 = X' * X

A2 =

   2   1   1
   1   4   1
   1   1   2



Apoi, pentru calcularea matricei S, vom avea nevoie doar de valorile proprii diferite de 0, deci fom lua matricea A de dimensiune mica, adica A<sub>2</sub>.

Vom folosi, pentru usurinta, functia <u>eig</u> pentru calcularea valorilor proprii:

In [8]:
%notam cu lambda vectorul coloana care ne da toate valorile proprii
%trebuie sa fim atenti si sa nu confundam cu vectorii specifici valorilor proprii
lambda = eig(A2)

lambda =

   1.00000
   2.00000
   5.00000



Observam ca valorile proprii rezultate din aceasta functie sunt mereu in ordine crescatoare. Noi formam matricea S cu elementele in ordine inversa:

In [9]:
S = diag(lambda(end:-1:1)) .^ (1/2)

S =

   2.23607   0.00000   0.00000
   0.00000   1.41421   0.00000
   0.00000   0.00000   1.00000



In caz ca pare dubios one-liner ul de mai sus, il vom explica pe pasi:
1. end reprezinta sfarsitul unui vector; deci, end:-1:1 inseamna elementele de la sfarsit la inceput(-1 este pasul), adica veectorul inversat
2. functia diag transforma un vector intr-o matrice diagonala
3. folosim operatorul special " .^ ", specific matlab/octave, care ridica fiecare element in parte la puterea 1/2.

In [10]:
%vom afisa fiecare etapa, pentru a fi totul mai evident
lambda
lambda(end:-1:1)
diag(lambda(end:-1:1))
diag(lambda(end:-1:1)) .^ (1/2)

diag(lambda(end:-1:1)) ^ (1/2)

printf("Hmmm... rezultatul este acelasi?!\n");

lambda =

   1.00000
   2.00000
   5.00000

ans =

   5.00000
   2.00000
   1.00000

ans =

Diagonal Matrix

   5.00000         0         0
         0   2.00000         0
         0         0   1.00000

ans =

   2.23607   0.00000   0.00000
   0.00000   1.41421   0.00000
   0.00000   0.00000   1.00000

ans =

Diagonal Matrix

   2.23607         0         0
         0   1.41421         0
         0         0   1.00000

Hmmm... rezultatul este acelasi?!


Ca un mic reminder, matricea S este matrice DIAGONALA. O proprietate importanta a acesteia este ca matricea ridicata la o putere q(numar real) este egala cu fiecare element ridicat la respectiva putere.
$$D = \begin{pmatrix} a_{11} & 0 & \dots & 0\\ 0 & a_{22} & \dots & 0 \\ \vdots & \vdots & \ddots & \vdots \\ 0 & 0 & \dots & a_{nn} \end{pmatrix}$$

Deci, matricea diagonala ridicata la o putere q, nu neaparat naturala, este:
$$ D^q = \begin{pmatrix} a_{11}^q & 0 & \dots & 0\\ 0 & a_{22}^q & \dots & 0 \\ \vdots & \vdots & \ddots & \vdots \\ 0 & 0 & \dots & a_{nn}^q \end{pmatrix}$$

Asadar, pentru o matrice diagonala, D^q <=> D.^q

Dupa ce am calculat valorile PROPRII necesare, prin urmare si gasirea matricei S, urmeaza sa calculam matricile U, respectiv V.
Pentru ca putem sa folosim mai multi algoritmi de calculare a vectorilor de valori proprii, ne vom folosi functia deja definita in matlab/octave, eig, care primeste ca parametru o matrice si returneaza vectorul de valori proprii in cazul in care dam un singur parametru, dar si MATRICEA DE VECTORI PROPRII, in cazul in care mai dam un parametru in plus.

<b>CALCULAREA LUI U:</b>

Prin urmare, daca ne folosim de functia deja data, eig, cu doi parametri, vom avea urmatoarele:

In [11]:
[U lambda] = eig(A1)

U =

   3.7558e-01  -2.4277e-01   3.4694e-17   8.1650e-01   3.6515e-01
   7.5943e-01   3.5108e-01  -1.9169e-16  -4.0825e-01   3.6515e-01
  -3.7558e-01   2.4277e-01  -7.0711e-01   7.1132e-17   5.4772e-01
  -8.2664e-03  -8.3662e-01   1.2100e-16  -4.0825e-01   3.6515e-01
  -3.7558e-01   2.4277e-01   7.0711e-01  -1.7106e-17   5.4772e-01

lambda =

Diagonal Matrix

  -1.6930e-16            0            0            0            0
            0   4.0523e-16            0            0            0
            0            0   1.0000e+00            0            0
            0            0            0   2.0000e+00            0
            0            0            0            0   5.0000e+00



Cateva observatii evidente: vectorul de valori proprii se transforma intr-o matrice diagonala, cu valorile sale insa in ordine crescatoare.
Prin urmare, ceea ce avem noi de facut este sa interschimbam coloanele din U, in felul urmator: prima coloana cu ultima coloana, a doua cu penultima, etc.

In [12]:
U = U(1:end, end:-1:1)

U =

   3.6515e-01   8.1650e-01   3.4694e-17  -2.4277e-01   3.7558e-01
   3.6515e-01  -4.0825e-01  -1.9169e-16   3.5108e-01   7.5943e-01
   5.4772e-01   7.1132e-17  -7.0711e-01   2.4277e-01  -3.7558e-01
   3.6515e-01  -4.0825e-01   1.2100e-16  -8.3662e-01  -8.2664e-03
   5.4772e-01  -1.7106e-17   7.0711e-01   2.4277e-01  -3.7558e-01



Poate ca vedem niste cifre si litere pe care nu prea le intelegem. Nu e nimic bizar in spate. 3.6515e-01 inseamna numarul 3.6515 este inmultit cu e-01, adica 10<sup>-1</sup>. Prin urmare, numerele care se termina in e-15, e-16, etc. le putem considera ca fiind egale cu 0, intrucat sunt mult prea mici pentru a fi considerate semnificative. Observam deci ca realitatea difera putin de fenomenul matematic, totul realizandu-se prin aproximari succesive, a caror eroare se propaga in timp, din pacate :( Ceea ce trebuie sa urmarim noi in general la Metode Numerice, si in general, in Computer Sience, este ca eroarea noastra sa se propage cat mai putin, pentru a putea avea un fenomen comparabil cu cel matematic.

In cazul de mai sus, nu trebuie sa ne facem probleme de genul, deci ceea ce am facut nu trebuie sa sperie pe nimeni. Prin urmare, putem face si comparatie: 3.6515e-01 este aproximativ egal cu 0.36515, adica chiar $\frac{\sqrt{30}}{15}$, un rezultat care ne bucura.

O observatie foarte IMPORTANTA!!! Vedem ca ultimele doua coloane din matrice difera enorm de mult fata de rezultatul matematic. Daca ne amintim bine, ultimele doua valori proprii erau egale cu 0, prin urmare am fi obtinut doi vectori liniar independenti.

Pentru a calcula un vector propriu, trebuie sa inlocuim in relatia $A \cdot v = \lambda \cdot v$ pe $\lambda$ cu valoarea proprie corespunzatoare. In cazul nostru, cu 0.

Avand doua valori proprii egale cu 0, atunci vom obtine din start doi vectori proprii.

Fie $$v = \begin{pmatrix} v_1 \\ v_2 \\ v_3 \\ v_4 \\ v_5 \end{pmatrix}$$

Din $A_1 \cdot v = 0$ (intrucat matricea noastra era de fapt A<sub>1</sub>),vom obtine urmatorul sistem de ecuatii:

$$\begin{cases} 2v_1 + v_3 + v_5 = 0 \\ v_2 + v_3 + v_4 + v_5 = 0 \\ v_1 + v_2 + 2v_3 + v_4 + v_5 = 0 \\ v_2 + v_3 + v_4 + v_5 = 0 \\ v_1 + v_2 + v_3 + v_4 + 2v_5 = 0 \end{cases}$$

Sistemul de mai sus este dublu nedeterminat, in care vom lua v<sub>4</sub> si v<sub>5</sub> termenii secundari. Prin urmare, noul nostru v va fi de forma:

$$v = \begin{pmatrix} -v_5 \\ -2v_5 - v_4 \\ v_5 \\ v_4 \\ v_5 \end{pmatrix} = v_4 \cdot \begin{pmatrix} 0 \\ -1 \\ 0 \\ 1 \\ 0 \end{pmatrix} + v_5 \cdot \begin{pmatrix} -1 \\ -2 \\ 1 \\ 0 \\ 1 \end{pmatrix}$$

Ce inseamna asta?? Ceea ce am obtinut mai sus, in cifre, este o baza, in care cei doi vectori sunt lininar independenti. Dand valori pentru v<sub>4</sub> si v<sub>5</sub>, vom obtine valori diferite pentru v, care este un vector propriu. Prin urmare noi am dat doua randuri de valori, adica 1 si 0, respectiv 0 si 1. Asta pentru a obtine cele mai "frumoase" rezultate. Pe hartie. Insa, daca am da si alte valori pentru v<sub>4</sub> si v<sub>5</sub>, am obtine valori diferite pentru v. Dar asta nu este o problema, intrucat nu conteaza, atata vreme cat v nu este altceva decat o combinatie liniara dintre elementele bazei ortogonale.

De aici rezulta ca si semnele schimbate pe care le tot vedem (in unele cazuri) nu reprezinta probleme computationale.

Prin urmare, unicitatea factorizarii SVD nu trebuie tratata mot a mot; unicitatea consta in existenta unei baze ortogonale unice, din care se pot desprinde diferite valori pentru matricile U, S, respectiv V. Rezultatul va fi mereu acelasi, adica matricea X.

<b>CALCULAREA LUI V:</b>

Procedeul este identic cu cel descris mai sus, in cazul lui U. Prin urmare vom calcula direct.

In [5]:
[V, lambda] = eig(A2)
V = V(1:end, end:-1:1)

error: 'A2' undefined near line 1 column 19
error: invalid use of end


Avand in vedere ca am reusit sa obtinem toate cele 3 matrici necesare, vom face si verificarea finala:

In [15]:
U * S * V'

error: operator *: nonconformant arguments (op1 is 5x5, op2 is 3x3)


Observam o mica eroare, datorata formei lui S, care este, momentan, 3 x 3. Vom rezolva elegant problema

In [16]:
%aflam dimensiunea lui A
[m n] = size(X);
%lasam octave sa faca autocomplete cu 0, in felul urmator
S(m,n) = 0
%deci atribuim unui element de pe o pozitie inexistenta 0, in felul urmator, se vor genera
%liniile si coloanele inexistente pana in acel loc, initializandu-se totodata cu 0

S =

   2.23607   0.00000   0.00000
   0.00000   1.41421   0.00000
   0.00000   0.00000   1.00000
   0.00000   0.00000   0.00000
   0.00000   0.00000   0.00000



In [17]:
%verificarea acum ar trebui sa mearga
U * S * V'

ans =

   1.0000e+00  -1.0448e-15   1.0000e+00
  -1.1542e-16   1.0000e+00   6.2552e-16
   3.5668e-16   1.0000e+00   1.0000e+00
  -2.2518e-17   1.0000e+00   2.7622e-16
   1.0000e+00   1.0000e+00   2.9723e-16



Aducandu-ne aminte de discutia cu e, putem constata cu mandrie ca rezultatul de mai sus este cel corect.

Putem face si o comparatie cu functia din octave, svd:

In [18]:
[S_nU_new ew V_new] = svd(X)
U
S
V

U_new =

  -3.6515e-01   8.1650e-01   2.3955e-16   1.1839e-01  -4.3126e-01
  -3.6515e-01  -4.0825e-01  -6.7318e-17  -5.6349e-01  -6.1845e-01
  -5.4772e-01  -1.0566e-16   7.0711e-01  -1.1839e-01   4.3126e-01
  -3.6515e-01  -4.0825e-01  -7.9046e-17   8.0027e-01  -2.4407e-01
  -5.4772e-01   2.6362e-16  -7.0711e-01  -1.1839e-01   4.3126e-01

S_new =

Diagonal Matrix

   2.2361        0        0
        0   1.4142        0
        0        0   1.0000
        0        0        0
        0        0        0

V_new =

  -4.0825e-01   5.7735e-01  -7.0711e-01
  -8.1650e-01  -5.7735e-01  -3.0175e-18
  -4.0825e-01   5.7735e-01   7.0711e-01

U =

   3.6515e-01   8.1650e-01   3.4694e-17  -2.4277e-01   3.7558e-01
   3.6515e-01  -4.0825e-01  -1.9169e-16   3.5108e-01   7.5943e-01
   5.4772e-01   7.1132e-17  -7.0711e-01   2.4277e-01  -3.7558e-01
   3.6515e-01  -4.0825e-01   1.2100e-16  -8.3662e-01  -8.2664e-03
   5.4772e-01  -1.7106e-17   7.0711e-01   2.4277e-01  -3.7558e-01

S =

   2.23607   0.00000  

Cu exceptia a catorva semne, se observa ca singurele lucruri care difera in mod deosebit sunt cele doua coloane de la sfarsitul lui U.
Motivul este clar, intrucat cei doi vectori vin in urma unui sistem dublu nedeterminat(dand posibilitatea mai multor valori care sa genereze cei doi vectori proprii).

Functia din octave este putin diferita fata de ce am facut noi aici, apeland la niste artificii computationale, pentru a putea fi mai eficienta. Noi am dorit sa intelegem conceptul, asa ca nu ne-a interesat in mod deosebit eficienta.

Ca tema de casa, veti avea de implementat, cap coada, o functie octave care sa realizeze factorizarea SVD

In [19]:
function [U S V] = SingValDec(X)
%X - matricea asupra careia dorim sa aplicam factorizarea SVD
%U - matrice patratica ce contine vectorii singulari de stanga
%S - matrice diagonala, ce are ca elemente valorile singulare
%V - matrice patratica ce contine vectorii singulari de dreapta

endfunction